File: programs/zuvi/src/instructions/initialize.rs
use anchor_lang::prelude::*;
use crate::{constants::*, errors::*, state::*};

/// 初始化系統配置
pub fn initialize(
    ctx: Context<Initialize>,
    api_signer: Pubkey,
    arbitrator: Pubkey,
    fee_receiver: Pubkey,
    usdc_mint: Pubkey,
    fee_rate: u16,
) -> Result<()> {
    // 檢查費率是否合理
    require!(
        fee_rate <= MAX_FEE_RATE,
        ZuviError::InvalidFeeRate
    );

    let config = &mut ctx.accounts.config;
    
    // 設定系統配置
    config.api_signer = api_signer;
    config.arbitrator = arbitrator;
    config.fee_receiver = fee_receiver;
    config.usdc_mint = usdc_mint;
    config.fee_rate = fee_rate;
    config.initialized = true;

    msg!("系統初始化成功");
    msg!("API Signer: {}", api_signer);
    msg!("仲裁者: {}", arbitrator);
    msg!("費用接收者: {}", fee_receiver);
    msg!("USDC Mint: {}", usdc_mint);
    msg!("費率: {} basis points", fee_rate);

    Ok(())
}

#[derive(Accounts)]
pub struct Initialize<'info> {
    /// 系統配置帳戶
    #[account(
        init,
        payer = authority,
        space = CONFIG_SIZE,
        seeds = [CONFIG_SEED],
        bump
    )]
    pub config: Account<'info, Config>,
    
    /// 初始化授權者（支付者）
    #[account(mut)]
    pub authority: Signer<'info>,
    
    /// 系統程式
    pub system_program: Program<'info, System>,
}

File: tests/initialize.ts
import * as anchor from "@coral-xyz/anchor";
import { Program } from "@coral-xyz/anchor";
import { Zuvi } from "../target/types/zuvi";
import { PublicKey, Keypair, SystemProgram } from "@solana/web3.js";
import bs58 from "bs58";

async function initialize() {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);

  const program = anchor.workspace.Zuvi as Program<Zuvi>;
  
  // API Signer
  const apiSignerKeypair = Keypair.fromSecretKey(
    bs58.decode("4K2Qa6nigxoESygxATeCX45L1HfPg4FoWau696CwvbX8KNCsNo4forbwiipHTXmiyEt6vzGXdghm5UUnREFJ2AGs")
  );
  
  // USDC Mint
  const usdcMint = new PublicKey("Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr");
  
  // 費率 0.4% = 40 basis points
  const feeRate = 40;
  
  // 衍生 Config PDA
  const [configPda] = PublicKey.findProgramAddressSync(
    [Buffer.from("config")],
    program.programId
  );
  
  // 檢查是否已初始化
  try {
    const config = await program.account.config.fetch(configPda);
    console.log("系統已初始化");
    console.log("Config:", {
      apiSigner: config.apiSigner.toString(),
      arbitrator: config.arbitrator.toString(),
      feeReceiver: config.feeReceiver.toString(),
      usdcMint: config.usdcMint.toString(),
      feeRate: config.feeRate,
      initialized: config.initialized
    });
    return;
  } catch {
    console.log("系統未初始化，開始初始化...");
  }
  
  try {
    const tx = await program.methods
      .initialize(
        apiSignerKeypair.publicKey,
        apiSignerKeypair.publicKey,  // 仲裁者也用同一個
        apiSignerKeypair.publicKey,  // 費用接收者也用同一個
        usdcMint,
        feeRate
      )
      .accountsStrict({
        config: configPda,
        authority: provider.wallet.publicKey,
        systemProgram: SystemProgram.programId,
      })
      .rpc();
    
    console.log("初始化成功");
    console.log("Transaction:", tx);
    console.log("Config PDA:", configPda.toString());
    
    // 驗證初始化
    const config = await program.account.config.fetch(configPda);
    console.log("Config:", {
      apiSigner: config.apiSigner.toString(),
      arbitrator: config.arbitrator.toString(),
      feeReceiver: config.feeReceiver.toString(),
      usdcMint: config.usdcMint.toString(),
      feeRate: config.feeRate,
      initialized: config.initialized
    });
    
  } catch (error) {
    console.error("初始化失敗:", error);
  }
}

initialize();